LAI with different Jmax for shaded and sunlit leaves

rcnijzink opened the pull request:
The multiplication of the fluxes is now moved to the end, in vom_add_hourly.
Some extra changes were needed to make sure everything works, for example for root water uptake.

There is now also a distinction between shaded and sunlit jmax25 (so four parameters to optimize now), and eventually also jact.

rcnijzink committed:
removed o_cat and caig from jact equations

rcnijzink committed:
correct dimensions of gstomg

rcnijzink committed:
correct dimensions of transpg

rcnijzink committed:
correct dimensions of etmg

rcnijzink committed:
corrected rwu limitations for et

rcnijzink committed:
moved multiplications with o_cait and caig to add_hourly

rcnijzink committed:
changed vom_tissue_water_et with multiplicaiton of etmt and o_cait

rcnijzink committed:
changed vom_subhourly with multiplicaiton of etmt and o_cait

rcnijzink committed:
changed vom_mqss with multiplicaiton of etmt and o_cait

rcnijzink committed:
small corrections

rcnijzink committed:
numerical changes in dmqt; adjust rootlim

rcnijzink committed:
moved multiplication with frac_sun and frac_shade to the end, different jmax_shade and jmax_sun

rcnijzink committed:
debugged initial attempt

rcnijzink committed:
initialized jmax25gs and jmax25ts

rcnijzink committed:
added all cases LAI

rcnijzink committed:
working on rootlim

rcnijzink committed:
debugging, errors in etmt__ + etmts__, setting daily assimilations back to zero

rcnijzink committed:
added some extra outputs

rcnijzink committed:
changed output header

rcnijzink committed:
updated documentation

rcnijzink committed:
added shaded parameters to OMP parallel

rcnijzink committed:
added lai to omp sce

rcnijzink committed:
small changes for stability in variable timestepping

rcnijzink committed:
correct foliage turnover costs

rcnijzink committed:
corrected optim cai_g determination

rcnijzink committed:
added dimension to etmg, transpg, gstomg, rootlim for caig; else rootlim does not work

rcnijzink committed:
small edits in sce

rcnijzink committed:
edited rlt outputs

rcnijzink committed:
corrected Respiration

rcnijzink committed:
edited cond3

rcnijzink committed:
cleaned up code

rcnijzink committed:
set shaded Rl and Jmax to zero

rcnijzink committed:
small corrections: uncessary evaluations; corrected jmax25ts in daily optim

rcnijzink committed:
added jact to output

rcnijzink committed:
corrected write_daily"

rcnijzink committed:
added jact to hourly outputs

rcnijzink committed:
set jmax25 minimum to 0

rcnijzink committed:
corrected output asst

rcnijzink committed:
adding extra outputs

rcnijzink committed:
corrected outputs

rcnijzink committed:
corrected calculation of diffuse radiation, with spitters et al, 1986

rcnijzink committed:
corrected calculation of diffuse radiation

rcnijzink committed:
corrected i_tcft

rcnijzink committed:
corrected unit of tree water storage from kg/m2 to g/m2

rcnijzink committed:
added pardir, diff and et to daily outputs

schymans commented on VOM_Fortran/VOM-code/modules.f90:
```diff
@@ -372,42 +375,58 @@ module vegmod
-
-      REAL*8  :: rlt_h(3,3)             ! Tree leaf respiration for different values of Jmax (rlt_h(2) is actual value) (mol/h)
+      REAL*8  :: gstomts                ! Shaded tree stomatal conductance (mol/m2/s)      
+      REAL*8  :: gstomg(3,3,3)            ! Grass stomatal conductance (mol/m2/s)
```
Is this then sunlit stomatal conductance? Would be good to specify for all variables where there is a shaded counterpart now.

rcnijzink replied:
Yes, it is, but in case the VOM is run without the shaded and sunlit leaves (it is an option), it is this gstom that does everything. I'll change it to !(Sunlit) grass stomatal conductance

schymans replied:
That would be great, also for the other variables that refer to sunlit leaves.

schymans commented on VOM_Fortran/VOM-code/modules.f90:
```diff
@@ -372,42 +375,58 @@ module vegmod
       REAL*8  :: rlt_d                  ! Daily tree leaf respiration (mol/d)
       REAL*8  :: rlt_y                  ! Annual tree leaf respiration (mol/y)
-      REAL*8  :: rlg_h(3,3,3)           ! Grass leaf respiration (mol/h)
+      REAL*8  :: rlg_h(3)           ! Grass leaf respiration (mol/h)
```
Why does rlg_h now only have dimension 3, not 3,3,3? I thought that with sunlit and shaded distinctions, there should be more, not fewer dimensions.

rcnijzink replied:
Previously, it was using three daily optimized values: jmax25g, grass cover, and LAI. The mutliplication with grass cover is now at the end. The LAI previously led to a fraction of absorbed radiation, which was also part of the Rl calculations, but we discussed this some time ago and concluded it shouldn't be there. It used to be this:

```
rlg_h(1,:,ii) = ((ca_h(th_) - gammastar) * pcg_d(1) * Ma_lg(ii) * jmaxg_h(:)    &
        &           * i_rlratio) / (4.d0 * (ca_h(th_) + 2.d0 * gammastar) &
        &           * (1.d0 + i_rlratio))  ! (3.24), (Out[312])
```
where Ma_lg was the fraction of absorbed radiation as a function of LAI. Now it is just this:

```
!    * respiration grasses
         rlg_h(:) = ((ca_h(th_) - gammastar) * jmaxg_h(:)    &
        &           * i_rlratio) / (4.d0 * (ca_h(th_) + 2.d0 * gammastar) &
        &           * (1.d0 + i_rlratio))  ! (3.24), (Out[312])
```

So now there is just the dimension 3, just due to the optimization of jmax25. There is now also a shaded leaf respiration, which I named rlgs_h (additional s).

schymans commented on VOM_Fortran/VOM-code/modules.f90:
```diff
@@ -416,25 +435,27 @@ module vegmod
-      REAL*8  :: q_mqx                  ! Tree maximum water content per ground area (kg/m2)
-      REAL*8  :: mqsst_                 ! Tree water content at steady state (kg/m2)
-      REAL*8  :: mqsstmin               ! Tree water content at turgor loss point (kg/m2)
+      REAL*8  :: mqt_                   ! Tree water content (g/m2)
```
Why the shift from kg to g? The VOM uses basic SI units throughout, so this can introduce confusion.

schymans commented on VOM_Fortran/VOM-code/modules.f90:
```diff
@@ -506,8 +528,8 @@ module vegmod
-      REAL*8  :: i_mqxtf                ! Total water storage capacity in living tissues of trees per unit md (kg/m^2)
+      REAL*8  :: i_mdtf                 ! Total dry mass of living tissues of trees per unit pc (g/m^2)
+      REAL*8  :: i_mqxtf                ! Total water storage capacity in living tissues of trees per unit md (g/m^2)
       REAL*8  :: i_rrootm               ! Root water uptake resistivity in soil (s)
```
Again, why g instead of kg?

schymans commented on VOM_Fortran/VOM-docu/source/outputfiles.rst:
```diff
@@ -35,9 +35,15 @@ The file "results_daily.txt" is an output file from VOM, containing daily weathe
-|rl               | Daily leaf respiration rate by trees and grasses combined (in mol m-2 d-1)                                  |
+|rlt              | Daily leaf respiration rate by trees (in mol m-2 d-1)                                                       |
++-----------------+-------------------------------------------------------------------------------------------------------------+
+|rlh              | Daily leaf respiration rate by grasses combined (in mol m-2 d-1)                                            |
```
What do you mean by "grasses combined"?

rcnijzink replied:
Oh that is a left over of the old sentence.... will correct!

schymans commented on VOM_Fortran/VOM-docu/source/vom_namelist.rst:
```diff
@@ -27,7 +27,7 @@ The file "vom_namelist" is an input file for the VOM, containing all settings fo
 +-----------------+----------------------------------------------------------------------------------------------------+-----------+
 |i_mdtf           | total dry mass of living tissues of trees per unit pc                                              | g/m^2     |
```
Why g not kg?

schymans commented on VOM_Fortran/VOM-docu/source/vom_namelist.rst:
```diff
@@ -27,7 +27,7 @@ The file "vom_namelist" is an input file for the VOM, containing all settings fo
 |i_mdtf           | total dry mass of living tissues of trees per unit pc                                              | g/m^2     |
 +-----------------+----------------------------------------------------------------------------------------------------+-----------+
-|i_mqxtf          | total water storage capacity in living tissues of trees per unit md                                | kg/m^2    |
+|i_mqxtf          | total water storage capacity in living tissues of trees per unit md                                | -         |
```
Is it really dimensionless? Then it should be "Relative", not "total".

rcnijzink replied:
Yes, I think so. Else the units do not match here:

```
!     * Set vegetation parameters

      q_md   = o_cait * i_mdtf + o_mdstore !dry matter (kg/m2) 
      q_mqx  = q_md * i_mqxtf              !dry matter (kg/m2) * water storage cap. living matter  (-) = (kg/m2)
```

schymans replied:
My bad, it says "total ... per unit md", so it is kg/kg.

schymans commented on VOM_Fortran/VOM-docu/source/vom_namelist.rst:
```diff
@@ -87,7 +88,7 @@ The file "vom_namelist" is an input file for the VOM, containing all settings fo
 |o_rtdepth        | tree rooting depth                                                                                 | m         |
 +-----------------+----------------------------------------------------------------------------------------------------+-----------+
-|o_mdstore        | wood water storage parameter of trees (can be in shufflepar)                                       | kg/m^2    |
+|o_mdstore        | wood water storage parameter of trees (can be in shufflepar)                                       | g/m^2     |
```
why g?

schymans commented on VOM_Fortran/VOM-input/vom_namelist:
```diff
@@ -11,7 +11,7 @@ i_toptf=0.0d0,          ! parameter to calculate adaptation of topt (range 0-1,
 i_toptstart=305.0d0,    ! start parameter for topt to calculate jmax(temp in K)
 i_rlratio=0.07d0,       ! leaf respiration parameter (rl=rlratio*assmax)
 i_mdtf=10000.0d0,       ! total dry mass of living tissues of trees per unit pc g/m^2
```
perhaps this should be changed to kg m-2

rcnijzink replied:
Yup, that's the culprit. Found that the VOM was using grams instead of kilograms. I didn't change any calculations, just the descriptions of the parameters in the commented part. But will change it to kg consistently.

schymans commented on VOM_Fortran/VOM-input/vom_namelist:
```diff
@@ -42,7 +42,7 @@ o_lambdatf=0.160181d4,  ! factor for calculating lambdat_d
 o_cait=0.3d0,           ! projected cover perennial vegetation (0-1)
 o_rtdepth=0.3d1,        ! tree rooting depth (m)
-o_mdstore=1.0d2,        ! wood water storage parameter of trees (can be in shufflepar)
+o_mdstore=1.0d2,        ! wood water storage parameter of trees (can be in shufflepar) (g/m^2)
```
Again, I would prefer kg instead of g

schymans commented on VOM_Fortran/VOM_test/ref_test/vom_namelist:
```diff
@@ -42,7 +42,7 @@ o_lambdatf=0.160181d4,  ! factor for calculating lambdat_d
 o_cait=0.3d0,            ! projected cover perennial vegetation (0-1)
 o_rtdepth=0.3d1,        ! tree rooting depth (m)
-o_mdstore=1.0d2,        ! wood water storage parameter of trees (can be in shufflepar)
+o_mdstore=1.0d2,        ! wood water storage parameter of trees (can be in shufflepar) (g/m^2)
```
Better kg, but need to check/adapt equation.

schymans reviewed (changes requested):
Great work, but the variable definitions should be made clearer and the units more consistent. Why not stick with kg throughout instead of g in parts of the code? It would also be good to describe the changes a bit in the docu and the merge request.

rcnijzink committed:
small changes due to PR; corrected g to kg, variable definitions

rcnijzink committed:
corrected grams to kilograms

rcnijzink committed:
changed documentation

rcnijzink committed:
changed documentation

rcnijzink commented:
I made the requested changes, and added a part in the documentation about the LAI dynamics. 

In fact, this PR is including the change in where the multiplication with o_cait and caig occurs:

- This happens now in vom_add_hourly:

```
        asst_h(:,:) = asst_h(:,:) + asst__(:,:) * dt_ * o_cait  
        assts_h(:,:) = assts_h(:,:) + assts__(:,:) * dt_ * o_cait
```


```
        etmt_h      = etmt_h      + dt_ * (etmt__ + etmts__) * o_cait
        etmg_h      = etmg_h      + dt_ * (etmg__(2,2,2) + etmgs__(2,2,2) ) * caig_d(2)
```

Another change is the modelling of dynamic LAI with shaded and sunlit leaves:

- There are shaded versions of the following parameters:
  jmax25t_d - jmax25ts_d
  jmax25g_d - jmax25gs_d
  
  jmaxt_h - jmaxts_h
  jmaxg_h - jmaxgs_h
  
  jactt - jactts
  jactg - jactgs
  
  etmt__ - etmts__
  etmg__ - etmgs__
 
  asst_h - assts_h
  assg_h - assgs_h
  
  asst_d - assts_d
  assg_d - assgs_d

  rlt_h - rlts_h
  rlg_h - rltg_h

- in vom_calc_derived there is a calculation of pardir and pardiff, first daily based on Roderick, 1999 and then hourly based on Spitters et al.,1986.

- Changed grams to kg in the VOM. There were some conversions in the code that had to be adjusted (times 1000).

schymans reviewed (approved):
Thank you, this is great!
